pkg/genericapiserver/server/serve.go

```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	max int
+}
+
+func (i *ipBasedLimit) incrementIp(ipser string) error {
+	if i.max <= 0 {
+		return nil
+	}
+	ip := net.ParseIP(ipser)
+	if ip == nil {
+		glog.Infof("Received request from client with invalid ip %s", ipser)
+		return limitedError{ip: ipser}
+	}
+	if ip.IsLoopback() {
+		return nil
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	current := i.limits[ipser]
+	if current >= i.ipMax || i.total >= i.max {
+		return limitedError{ip: ipser}
+	}
+	i.total++
+	i.limits[ipser] = current + 1
+	return nil
+}
+
+func (i *ipBasedLimit) decrementIp(ipser string) {
+	if i.max <= 0 {
+		return
+	}
+	ip := net.ParseIP(ipser)
+	if ip == nil || ip.IsLoopback() {
+		return
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	i.total--
+	current := i.limits[ipser]
+	if current > 1 {
+		i.limits[ipser] = current - 1
+	} else {
+		glog.Infof("Cleaned up all references for IP %s", ipser)
+		delete(i.limits, ipser)
+	}
+}
+
+// The actual structure/handler which attaches the limiting code to a connection
+type ipBasedLimitHandler struct {
+	// The limit structure for this connection type.
+	ipLimit *ipBasedLimit
+}
+
+func (h ipBasedLimitHandler) init(conn *net.TCPConn) error {
+	return nil
+}
+
+func (h ipBasedLimitHandler) wrap(conn net.Conn) (net.Conn, error) {
+	ip, _, err := net.SplitHostPort(conn.RemoteAddr().String())
 	if err != nil {
+		conn.Close() // Not decrementing as we have not yet incremented
+		glog.Infof("Something went wrong getting the remote address %v", err)
 		return nil, err
 	}
-	tc.SetKeepAlive(true)
-	tc.SetKeepAlivePeriod(defaultKeepAlivePeriod)
-	return tc, nil
+	local := conn.LocalAddr().String()
+	glog.Infof("Entered ip based limit Accept() for IP %s on %s", ip, local)
+	err = h.ipLimit.incrementIp(ip)
+	if err != nil {
+		conn.Close() // Not decrementing as we did not successfully increment
+		glog.Infof("Denied accept to IP %s", ip)
+		return nil, err
+	}
+	glog.Infof("Allowed accept to IP %s", ip)
+	return &ipBasedLimitListenerConn{Conn: conn, ipLimit: h.ipLimit, ip: ip}, nil
+}
+
+// The structure which is responsible for cleaning up/decrementing the related ip limit
+type ipBasedLimitListenerConn struct {
+	// The underlying connection used to for the request/response.
+	net.Conn
+
+	// Makes sure we do not decrement more than once for the given connection
+	releaseOnce sync.Once
+
+	// Reference to the relevant IP Limiting structure.
+	ipLimit *ipBasedLimit
+
+	// The IP addess used when reserving/incrementIp so we can clean up
+	ip string
+}
+
+func (c *ipBasedLimitListenerConn) Close() error {
+	err := c.Conn.Close()
+	c.releaseOnce.Do(func() { c.ipLimit.decrementIp(c.ip) })
+	return err
+}
+
+// A custom error which tells http/server.go Serve() to retry Accept() quickly
+type limitedError struct {
+	// Present for debugging - IP which saw the problem
+	ip string
+}
+
+func (l limitedError) Error() string {
+	return "Accept() Exceeded maximum concurrent requests for IP " + l.ip
```

Comment:
Or exceeded the overall limit?

Go error messages traditionally start with lowercase letters and don't end with punctuation (so that they can be concatenated and not look weird if there are multiple ones). That's a long way of saying s/Exceeded/exceeded/.